torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -0,0 +1,411 @@
+
+  std::unique_lock<std::mutex> lock(mutex_);
+
+  // See if we already have a connection to this address or not
+  auto it = connectedPipes_.find(toWorkerInfo.id_);
+  if (it == connectedPipes_.end()) {
+    std::tie(it, std::ignore) = connectedPipes_.emplace(
+        toWorkerInfo.id_, ClientPipe(context_->connect(url)));
+  }
+  ClientPipe& clientPipe = it->second;
+  auto& pendingResponseMessage = clientPipe.pendingResponseMessage_;
+
+  std::shared_ptr<FutureMessage> futureResponseMessage =
+      std::make_shared<FutureMessage>();
+  requestMessage.setId(nextMessageID_++);
+  pendingResponseMessage[requestMessage.id()] = futureResponseMessage;
+
+  // Don't need to hold lock while calling tensorpipe API.
+  lock.unlock();
+
+  pipeWrite(
+      clientPipe.pipe_,
+      std::move(requestMessage),
+      [this, &clientPipe, futureResponseMessage](
+          const tensorpipe::Error& error) {
+        if (error) {
+          LOG(WARNING) << "client write error: " << error.what();
+          futureResponseMessage->setError(error.what());
+          return;
+        }
+
+        pipeRead(
+            clientPipe.pipe_,
+            [this, &clientPipe](
+                const tensorpipe::Error& error, Message&& responseMessage) {
+              if (error) {
+                LOG(WARNING) << "Read response error: " << error.what();
+                std::lock_guard<std::mutex> lock(mutex_);
+                // We may get garbage content in responseMessage upon error.
+                // Flushing all future messages belonging to this pipe due to
+                // error state.
+                for (auto& p : clientPipe.pendingResponseMessage_) {
+                  std::shared_ptr<FutureMessage>& futureMessage = p.second;
+                  futureMessage->setError(error.what());
+                }
+                clientPipe.pendingResponseMessage_.clear();
+                clientPipe.readError_ = true;
+                return;
+              }
+
+              // Identify future response message by message ID
+              uint64_t messageId = responseMessage.id();
+              std::shared_ptr<FutureMessage> futureResponseMessage;
+              {
+                std::lock_guard<std::mutex> lock(mutex_);
+                // A read error will lead all following callbacks to be
+                // invoked with error, and shouldn't reach here.
+                TORCH_INTERNAL_ASSERT(
+                    !clientPipe.readError_, "Shouldn't in error state");
+                auto it = clientPipe.pendingResponseMessage_.find(messageId);
+                TORCH_INTERNAL_ASSERT(
+                    it != clientPipe.pendingResponseMessage_.end(),
+                    "message ID ",
+                    messageId,
+                    " is not recognized");
+                futureResponseMessage = std::move(it->second);
+                clientPipe.pendingResponseMessage_.erase(it);
+              }
+
+              threadPool_.run(
```

Comment:
Maybe a noob q, but why is marking the future as complete/error deferred to a thread pool task? I don't think we're doing any I/O/async/heavy work here, so maybe it's best to just run this in line?

Reply:
Marking a future as complete can entail an unbounded amount of work/processing that doesn't need to be in an IO thread.

Reply:
Yes, and in fact in #39010 we make sure that all futures are completed from the thread pool (in this PR we had missed a couple).